added data packets and packet filter

B3n30 opened the pull request:
This is the filter algorithm for the beacon sender. Since sender mac will never be the broadcast mac setting filter to the broadcast mac, will give all the beacons stored in the deque

B3n30 committed:
added packet filter for beacon packets

B3n30 committed:
Added data packets and data_mutex

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -69,9 +69,21 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
+                }
+            }
+        }
+        return result_queue;
```
I suggest moving this to a lambda or even a real function so we can re-use it for the data filter later (also you can give it a more descriptive name which makes the code more readable)

B3n30 replied:
I thought "PopWifiPackets" is exactly that function...

Subv replied:
std::move(result_queue)

B3n30 replied:
"Moving a local object in a return statement prevents copy elision". Should i change it nevertheless

Subv replied:
Leave it, we want copy elision here, we can change it later if it produces sub-optimal code.

B3n30 committed:
Added data packets and FilterAndPopPackets lambda

Fixup: Added lambda for EmplaceBackAndCheckSize, changed lamdas to capture by ref and added space before else

Fixup: pointer -> reference, doubled emplace_back beacon, and clang

B3n30 commented:
Renamed the PR

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -61,31 +61,44 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
 std::deque<WifiPacket> RoomMember::PopWifiPackets(WifiPacket::PacketType type, const MacAddress& mac_address) {
-    if (type == WifiPacket::PacketType::Beacon) {
-        std::lock_guard<std::mutex> lock(beacon_mutex);
+    auto FilterAndPopPackets = [=](std::deque<WifiPacket>* source_queue) {
```
Capture and pass by ref?

B3n30 replied:
Addressed

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -61,31 +61,44 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
                     result_queue.emplace_back(std::move(*it));
-                    it = beacon_queue.erase(it);
+                    it = source_queue->erase(it);
                 }else {
```
nit: missing a space (while you are at it anyway)

B3n30 replied:
Addressed

JayFoxRox commented on src/room_member.cpp:
```diff
@@ -61,31 +61,44 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
+        // If we have more than the max number of buffered data packets, discard the oldest one
+        if (data_queue.size() > MaxDataQueueSize) {
+            data_queue.pop_front();
+        }
```
same thing: Probably want to use a lambda / function to avoid code duplication
(or: `auto& queue = ... ? beacon_queue : data_queue`)

B3n30 replied:
Went for the lambda...

Subv commented on src/room_member.cpp:
```diff
@@ -29,6 +29,16 @@ RoomMember::~RoomMember() {
 void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
+    WifiPacket wifi_packet{};
+    
+    auto EmplaceBackAndCheckSize = [&](std::deque<WifiPacket>* queue, size_t maxSize) {
```
max_size.
Lower snake case for variables

Subv replied:
Why a pointer instead of a reference?

B3n30 replied:
I'm somehow used to use pointers because of some stupid physics libraries...

Subv commented on src/room_member.cpp:
```diff
@@ -56,24 +65,39 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
     if (type == WifiPacket::PacketType::Beacon) {
         std::lock_guard<std::mutex> lock(beacon_mutex);
         beacon_queue.emplace_back(std::move(wifi_packet));
```
You're calling emplace_back twice for beacons

Subv commented on src/room_member.cpp:
```diff
@@ -56,24 +65,39 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
 }
 
+
 std::deque<WifiPacket> RoomMember::PopWifiPackets(WifiPacket::PacketType type, const MacAddress& mac_address) {
```
Throughout all the files, leave a space after each flow control keyword.

If (mac_address == NoPreferredMac)

Subv commented on src/room_member.cpp:
```diff
@@ -56,24 +65,39 @@ void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
 
+
 std::deque<WifiPacket> RoomMember::PopWifiPackets(WifiPacket::PacketType type, const MacAddress& mac_address) {
+    auto FilterAndPopPackets = [&](std::deque<WifiPacket>* source_queue) {
```
Why a pointer instead of a reference?

B3n30 commented:
Addressed everything except return std::move(result_queue);

Subv reviewed (approved):
lgtm, squash and merge
